### ISSUE TITLE
OOM Killed Allocations

### ISSUE THREAD

**Issue by mpass99** [openHPI/poseidon#397]
Related to #396

> Currently we observe some runners being OOM Killed (just) in the Poseidon logs (Lack of Error reporting or ok?). This leads to the runner being restarted and added to the idle runners again (even if it was used before the restart).

We expect that this leads to interfering runner management actions if the idle runner was re-assigned to a new user within the timeout of old executions.

 > I know that we might see an OOM killed message, when too much memory is consumed within a runner. One example to trigger this error might be the following: In an R script, just type the following:
> ``` R
> table(airquality)
> ```

The difference between the two types of `OOM Killed` is:
- For the known case (`table(airquality)`) the command is killed within the container.
  - The runner can be further used for the next allocation.
- For the new case the whole allocation/container is OOM Killed.
  - The runner is restarted and re-added to the prewarming pool.
  - Existing WebSocket connections timeout.
  - Currently not reproducible
  - `map[Allocation:map[ClientDescription:No tasks have started ClientStatus:pending DesiredStatus:run JobID:29-4dbd13a6-0acd-11ee-a4f7-fa163e079f19 NodeName:nomad-agent-terraform-1 Resources:map[CPU:20 MemoryMB:30 MemoryMaxMB:256 ] TaskStates:map[default-task:map[Events:[map[Details:map[exit_code:137 exit_message:OOM Killed oom_killed:true signal:0] Exit Message: "OOM Killed" Type:Terminated]]] ]]]]`

Expected behaviour would be:
> If this memory limit we defined is the root cause of those issues, the WebSocket connection to the browser should also be closed, but of course I don't know about the other WebSocket connection between Poseidon and CodeOcean. I have the feeling that in any case, we should also stop the WebSocket connection if a runner becomes unavailable (such in this case). Do you see any possibility to tackle this issue as well?

**Comment by mpass99** [comment]
According to the Telegraf monitoring the (memory of the) Nomad cluster was barely used in the time of the observed occurrences. This strikes out the possibility that the Linux kernel of the host killed the allocations due to low available memory.

**Comment by MrSerth** [comment]
Thanks for creating a dedicated issue and summarizing the findings. It's also interesting to see that the existing case (user program takes too much RAM, such as with `table(airquality)`) is handled appropriately, but this case obviously is different. 

With some further debugging, I saw that the following output was provided to a learner:

```
status: timeout
 stdout: waiting on pid 44: waiting on PID 44 in sandbox "aeb3d07ab278eab8024a3bb44c9bbaacac5d9eca870385db22a9ad41f68494cc": urpc method "containerManager.WaitPID" failed: EOF

 stderr: test_assess (r232.Assess) ... waiting on pid 48: waiting on PID 48 in sandbox "aeb3d07ab278eab8024a3bb44c9bbaacac5d9eca870385db22a9ad41f68494cc": urpc method "containerManager.WaitPID" failed: EOF
```

Based on this output, I probably was able to reproduce it. Check [this exercise `Python20 Aufgabe 2.3.2`](https://codeocean.openhpi.de/exercises/666/implement) with input:

```py
verbleibende_Zeit = 100
while verbleibende_Zeit >= 17 :
    print("Wir spielen noch eine Runde.")
print("Nun ist Zeit zum Schlafen.")
```

Of course, this code will create an infinite loop. While doing so, it will also spam the output. In a regular "run" on CodeOcean, this is not a major issue, everything behaves as expected (including the browser, which is slowing down due to the massive output). However, the error is triggered when **assessing** this solution. In this case, weird things happen, for which I don't have an explanation yet...

Does this help you to reproduce the issue?

**Comment by mpass99** [comment]
Thank you for this additional information! With it I was able to reproduce the error. It turns out that the program is just exceeding its memory limit. `The difference between the two types of OOM Killed is` the use/nonuse of gVisor. With gVisor enabled the whole container is OOM Killed, otherwise just the process.

The minimal reproducible example is `dd if=/dev/zero of=/dev/null bs=500M` or 
``` python
import os

os.system("dd if=/dev/zero of=/dev/null bs=500M")
```

I guess we have to accept allocations being `OOM Killed`, but we should not wait for the execution timeout to exceed when we already know that the allocation has been stopped. So far, I could think of the following solutions.
1. Abort the WebSocket connection to Nomad Poseidon-side aka Destroy runner when the allocation is stopped. Implemented in #401 
2. Abort the WebSocket connection to Nomad Nomad-side aka investigate why Nomad does not abort it

**Comment by MrSerth** [comment]
Awesome, thanks for identifying the cause for the two differences for OOM killed executions. It makes sense that gVisor is involved here, but it would have been better without it changing this behavior ;).

I also agree that we should not wait for the timeout to exceed before taking the appropriate actions. Therefore, I like the approach you chose and also think that, while Nomad should close the connection, it's also fine for Poseidon to clean up.
